wire-ios-data-model/Source/Core Crypto/CoreCryptoKeyProvider.swift

```diff
@@ -17,27 +17,64 @@
 //
 
 import Foundation
+import WireCoreCrypto
 import WireLogging
 import WireSystem
 
 public class CoreCryptoKeyProvider {
 
     public init() {}
 
-    public func coreCryptoKey(createIfNeeded: Bool) throws -> Data {
+    public func coreCryptoKey(
+        createIfNeeded: Bool,
+        path: String
+    ) async throws -> Data {
         removeLegacyKeyIfNeeded()
 
         do {
-            return try fetchCoreCryptoKey()
+            return try fetchCoreCryptoKeyV2()
         } catch {
             if createIfNeeded {
-                return try createCoreCryptoKey()
+                guard let oldKey = try? fetchCoreCryptoKey() else {
+                    return try createCoreCryptoKeyV2()
+                }
+                return try await migrateDatabaseKey(path: path, oldKey: oldKey)
             } else {
                 throw error
             }
         }
     }
 
+    private func fetchCoreCryptoKeyV2() throws -> Data {
+        let item = CoreCryptoKeychainItemV2()
+        let key: Data = try KeychainManager.fetchItem(item)
+        WireLogger.coreCrypto.info("Core crypto key_v2 exists: \(key.base64String()). Returning...")
+        return key
+    }
+
+    private func createCoreCryptoKeyV2() throws -> Data {
+        let item = CoreCryptoKeychainItemV2()
+        WireLogger.coreCrypto.info("Core crypto key_v2 doesn't exist. Creating...")
+        let key = try KeychainManager.generateKey(numberOfBytes: 32)
+        WireLogger.coreCrypto.info("Created core crypto key_v2: \(key.base64String()). Storing...")
+        try KeychainManager.storeItem(item, value: key)
+        WireLogger.coreCrypto.info("Stored core crypto key_v2. Returning...")
+        return key
+    }
+
+    private func migrateDatabaseKey(path: String, oldKey: Data) async throws -> Data {
```

Comment:
question: should we delete the old key?

Reply:
I think yes, I don't have ideas why we can use it again.